drivers: accel: Add ADXL371 driver

Brandon-Hurst opened the pull request:
## Pull Request Description

Add a driver for the ADXL371. This accelerometer is designed for impact / shock detection applications while utilizing very low power. 

Testing performed uses a project in an external build system which brings in No-OS as a submodule. This project will be released separately dependent on merging this PR. The relevant source files are attached.

[adxl371_noos_main.zip](https://github.com/user-attachments/files/22604714/adxl371_noos_main.zip)

The .rst document inside the driver folder documents how to utilize the shock detection features, as well as the recommended FIFO configurations for different use-cases for the device such as capturing single peaks, sequences of peaks, or complete peak events. Most of the relevant code in this document is taken directly from the example project. 

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have complied with the [Submission Checklist](http://analogdevicesinc.github.io/no-OS/contributing.html#submission-checklist)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

Brandon-Hurst committed:
accel: adxl371: Add ADXL371 driver

- SPI/I2C platform files
- Header / Source driver files for the ADXL371

Driver contains a "get_fifo_xyz_data_workaround" function.
This is to indicate that the function complies with the
workaround for the anomaly mentioned at the end of the datasheet.
This requires some extra guarantees by the user, which are
documented in the driver code.

Signed-off-by: Brandon Hurst <brandon.hurst@analog.com>

Brandon-Hurst committed:
accel: adxl371: Add RST file for ADXL371 driver

Signed-off-by: Brandon Hurst <brandon.hurst@analog.com>

amiclaus reviewed (commented):
please fix all the ci issues. also you should have only 2 commits. one fro the driver itself and one for the documentation.

amiclaus reviewed (commented):
please perform a rebase with the latest main branch, and check if the build fails.

Brandon-Hurst commented:
Did the rebase, thanks for the help! Should be good now.

amiclaus commented on drivers/accel/adxl371/adxl371.c:
```diff
@@ -0,0 +1,914 @@
+	int2_config = (ADXL371_INT2_MAP_DATA_RDY_MODE(int2.data_rdy) |
+		       ADXL371_INT2_MAP_FIFO_RDY_MODE(int2.fifo_rdy) |
+		       ADXL371_INT2_MAP_FIFO_FULL_MODE(int2.fifo_full) |
+		       ADXL371_INT2_MAP_FIFO_OVR_MODE(int1.fifo_ovr) |
```
this is wrong.

Brandon-Hurst replied:
Pulled this from the ADXL372 driver:

https://github.com/analogdevicesinc/no-OS/blob/ba04e29fae209b104fb0d9de878934740ce716e9/drivers/accel/adxl372/adxl372.c#L373-L410

Could you explain what the issue is here?

amiclaus commented on drivers/accel/adxl371/adxl371.c:
```diff
@@ -0,0 +1,914 @@
+{
+	int ret;
+
+	ret = no_os_spi_remove(dev->spi_desc);
```
when do we use i2c_remove?

amiclaus commented on drivers/accel/adxl371/adxl371.c:
```diff
@@ -0,0 +1,914 @@
+ * @param cnt - How many samples should be retrieved from the FIFO DATA reg
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t adxl371_get_fifo_xyz_data_workaround(struct adxl371_dev *dev,
```
return type should be int for all functions.

amiclaus reviewed (commented):
some first comments on my side. my suggestion is to take as reference some of the newly added drivers into our branch and follow that format.
also have a look at the commit format used.
https://analogdevicesinc.github.io/no-OS/ might help too.

RaduSabau1 commented on drivers/accel/adxl371/adxl371.c:
```diff
@@ -0,0 +1,914 @@
+	enum adxl371_op_mode op_mode = adxl371_get_op_mode(dev);
+
+	// @workaround Set the timing register to external sync along with odr
+	adxl371_set_external_sync(dev);
```
`ret` should be used for error checking, please check other occurrences

RaduSabau1 commented on drivers/accel/adxl371/adxl371.c:
```diff
@@ -0,0 +1,914 @@
+
+	if (dev_id != ADXL371_DEVID_VAL || part_id != ADXL371_PARTID_VAL) {
+		printf("failed to read id (0x%X : 0x%X)\n", dev_id, part_id);
+		ret = -1;
```
please use appropriate errno code, perhaps `-ENODEV`

RaduSabau1 reviewed (commented):
some comments from my side

Brandon-Hurst commented:
Made requested driver fixes and updated commit titles. I re-tested again with the original application, modified slightly to match the changes. 

[adxl371_noos_main_update.zip](https://github.com/user-attachments/files/25277950/adxl371_noos_main_update.zip)

The cppcheck CI still is complaining but it looks like it just can't reach a mirror for kernel.org:

```bash
--2026-02-13 01:32:31--  http://mirrors.kernel.org/ubuntu/pool/universe/t/tinyxml2/libtinyxml2-6a_7.0.0+dfsg-1build1_amd64.deb
Resolving mirrors.kernel.org (mirrors.kernel.org)... 142.0.200.124, 2605:f480:58:1:0:1994:3:14
Connecting to mirrors.kernel.org (mirrors.kernel.org)|142.0.200.124|:80... failed: Connection refused.
Connecting to mirrors.kernel.org (mirrors.kernel.org)|2605:f480:58:1:0:1994:3:14|:80... failed: Network is unreachable.
```

amiclaus reviewed (commented):
> Made requested driver fixes and updated commit titles. I re-tested again with the original application, modified slightly to match the changes.
> 
> [adxl371_noos_main_update.zip](https://github.com/user-attachments/files/25277950/adxl371_noos_main_update.zip)
> 
> The cppcheck CI still is complaining but it looks like it just can't reach a mirror for kernel.org:
> 
> ```shell
> --2026-02-13 01:32:31--  http://mirrors.kernel.org/ubuntu/pool/universe/t/tinyxml2/libtinyxml2-6a_7.0.0+dfsg-1build1_amd64.deb
> Resolving mirrors.kernel.org (mirrors.kernel.org)... 142.0.200.124, 2605:f480:58:1:0:1994:3:14
> Connecting to mirrors.kernel.org (mirrors.kernel.org)|142.0.200.124|:80... failed: Connection refused.
> Connecting to mirrors.kernel.org (mirrors.kernel.org)|2605:f480:58:1:0:1994:3:14|:80... failed: Network is unreachable.
> ```

That's an unrelated issue, so you can ignore it. But for the documentation build, the issues seem to be valid:
```
Generating caller graph for function adxl367_set_fifo_mod/home/vsts/work/1/s/drivers/accel/adxl371/adxl371.c:604: warning: The following parameter of adxl371_get_internal_sync(struct adxl371_dev *dev, bool *internal_sync) is not documented:
/home/vsts/work/1/s/drivers/accel/adxl371/adxl371.c:183: warning: The following parameter of adxl371_get_op_mode(struct adxl371_dev *dev, enum adxl371_op_mode *op_mode) is not documented:
/home/vsts/work/1/s/drivers/accel/adxl371/adxl371.c:604: warning: The following parameter of adxl371_get_internal_sync(struct adxl371_dev *dev, bool *internal_sync) is not documented:
/home/vsts/work/1/s/drivers/accel/adxl371/adxl371.c:183: warning: The following parameter of adxl371_get_op_mode(struct adxl371_dev *dev, enum adxl371_op_mode *op_mode) is not documented:
```

amiclaus commented on drivers/accel/adxl371/adxl371.c:
```diff
@@ -0,0 +1,979 @@
+	if (ret < 0)
+		goto error;
+
+	ret |= no_os_gpio_get(&dev->gpio_int2,
```
don't use `|=` treat each error independently. check the entire code.

amiclaus commented on drivers/accel/adxl371/adxl371.c:
```diff
@@ -0,0 +1,979 @@
+
+	*device = dev;
+	printf("adxl371 successfully initialized\n");
+	no_os_mdelay(1000);
```
1 second delay? add a comment if so high value is needed.

Brandon-Hurst replied:
Another thing from ADXL372:
https://github.com/analogdevicesinc/no-OS/blob/6e37229f4a2805ecb418d691283a1ba1774af397/drivers/accel/adxl372/adxl372.c#L831-L839

I see nothing in the datasheet that states this is necessary. I tested this without the delay in my example and it works fine. So I'll remove it.

amiclaus commented on drivers/accel/adxl371/README.rst:
```diff
@@ -0,0 +1,470 @@
+
+The ADXL371 driver facilitates communication and control through
+functions like ``adxl371_write_reg``, ``adxl371_read_reg``, and
+``adxl371_write_mask`` for manipulatibg register values.
```
typo.

amiclaus reviewed (commented):
some comments on my side.

stefpopa commented on drivers/accel/adxl371/adxl371_i2c.c:
```diff
@@ -0,0 +1,118 @@
+/*******************************************************************************
+ *   @file   adxl371_i2c.c
+ *   @brief  Implementation of adxl371 I2C Driver.
+ *   @author SPopa (stefan.popa@analog.com)
```
I believe the author is wrong here. As much as I would like to take credit for this driver, I think it's not fair :-)

stefpopa reviewed (commented):
I did not check, but how different is it from adx372? Could we add support for this part inside the same adxl372 driver? Feel free to disagree and argument against this idea.

Brandon-Hurst commented on drivers/accel/adxl371/adxl371_i2c.c:
```diff
@@ -0,0 +1,118 @@
+		return -EINVAL;
+
+	buf[0] = reg_addr;
+	memset(&buf[1], 0x00, count - 1);
```
This is another one from the adxl372_i2c driver. 
https://github.com/analogdevicesinc/no-OS/blob/6e37229f4a2805ecb418d691283a1ba1774af397/drivers/accel/adxl372/adxl372_i2c.c#L108-L109

I think the intent here is to zero-fill the array of tx/rx data. In the case of tx it is one byte but in the rx case it is pre-filling the entire array with 0s. Perhaps not strictly necessary, but can be helpful to distinguish real data if you were debugging, for example. If you'd like me to take it out, I can.

Brandon-Hurst commented:
> I did not check, but how different is it from adx372? Could we add support for this part inside the same adxl372 driver? Feel free to disagree and argument against this idea.

I tried this in the beginning, but found some issues as described in some of the comments. The registers are similar, but have some minor value differences in terms of BW, ODR, Wakeup rate, etc. 

The current ADXL372 driver also does not implement the workaround for the silicon anomaly at the end of the datasheet regarding the FIFO modes. This means that samples come in out of order in all FIFO modes, so an X will show up where a Z should be, etc. I'm happy to help add the changes to the ADXL372 driver as we should probably do that as well. 

Could probably be supported by the same driver with different header files / #ifdefs, but I guess it could get dicey in a project with many accelerometers (imagining you had to compensate with generic register macros e.g. ADXL_BW_REG).

amiclaus reviewed (commented):
> > I did not check, but how different is it from adx372? Could we add support for this part inside the same adxl372 driver? Feel free to disagree and argument against this idea.
> 
> I tried this in the beginning, but found some issues as described in some of the comments. The registers are similar, but have some minor value differences in terms of BW, ODR, Wakeup rate, etc.
> 
> The current ADXL372 driver also does not implement the workaround for the silicon anomaly at the end of the datasheet regarding the FIFO modes. This means that samples come in out of order in all FIFO modes, so an X will show up where a Z should be, etc. I'm happy to help add the changes to the ADXL372 driver as we should probably do that as well.
> 
> Could probably be supported by the same driver with different header files / #ifdefs, but I guess it could get dicey in a project with many accelerometers (imagining you had to compensate with generic register macros e.g. ADXL_BW_REG).

you can make some device specific info structure or add parameters in the device structure for the values that are different. also for the anomaly handling i suppose the `_workaround` function would be the replacement for the original get_xyz_fifo function from adxl372. there you can use function pointers :).

stefpopa commented:
> > > I did not check, but how different is it from adx372? Could we add support for this part inside the same adxl372 driver? Feel free to disagree and argument against this idea.
> > 
> > 
> > I tried this in the beginning, but found some issues as described in some of the comments. The registers are similar, but have some minor value differences in terms of BW, ODR, Wakeup rate, etc.
> > The current ADXL372 driver also does not implement the workaround for the silicon anomaly at the end of the datasheet regarding the FIFO modes. This means that samples come in out of order in all FIFO modes, so an X will show up where a Z should be, etc. I'm happy to help add the changes to the ADXL372 driver as we should probably do that as well.
> > Could probably be supported by the same driver with different header files / #ifdefs, but I guess it could get dicey in a project with many accelerometers (imagining you had to compensate with generic register macros e.g. ADXL_BW_REG).
> 
> you can make some device specific info structure or add parameters in the device structure for the values that are different. also for the anomaly handling i suppose the `_workaround` function would be the replacement for the original get_xyz_fifo function from adxl372. there you can use function pointers :).

If registers are similar, then we should probably add support as part of the same driver. That's the general procedure with other parts as well.

buha commented:
@Brandon-Hurst please reopen when you add the 371 as part of the adxl372 driver
